src/platform/PICO/adc_pico.c

```diff
@@ -0,0 +1,260 @@
+    if (channelCount == 0) {
+        /* don't enable the adc/dma - exit immediately */
+        return;
+    }
+
+    if (channelCount == 3) {
+        // padding required due to DMA ring buffer supporting 2^X
+        // enable the internal temp sensor to add one more reading.
+        adc_set_temp_sensor_enabled(true);
+        mask |= (1u << PICO_ADC_INTERNAL_TEMP_CHANNEL);
+    }
+
+    // empty the ADC fifo for safety
+    adc_fifo_drain();
+
+    // set the round robin for ADC
+    adc_set_round_robin(mask);
+
+    const int firstEnabled = llog2(mask & -mask);
+    adc_select_input(firstEnabled);
+
+    // Write each completed conversion to the sample FIFO
+    // Enable DMA data request (DREQ)
+    // Fire DREQ when FIFO has at least 1 sample
+    // Ignore ERR bit, save masking the result (as it is 12 bit).
+    // Don't shift each sample to 8 bits when pushing to FIFO
+    adc_fifo_setup(true, true, 1, false, false);
+
+    /*
+        Conversion requires 96 cycles per sample, at ADC clock of 48MHz.
+        clkdiv divides ADC clock to generate next auto trigger event
+
+        clkdiv = (48,000,000 / (sample_rate)) - 1
+
+        Sample rate required at 10hz for X channels = 10 x X
+
+        65535.f + 255.f / 256.f for as slow as possible
+        0 for as fast as possible
+    */
+    adc_set_clkdiv(65535.f + 255.f / 256.f);
+
+    const dmaIdentifier_e dma_id = dmaGetFreeIdentifier();
+    if (dma_id == DMA_NONE || !dmaAllocate(dma_id, OWNER_ADC, 0)) {
+        return; // No free DMA channel available
+    }
+    const uint8_t dmaChannel = DMA_IDENTIFIER_TO_CHANNEL(dma_id);
+
+    // --- DMA Setup ---
+    dma_channel_config cfg = dma_channel_get_default_config(dmaChannel);
+
+    // Configure DMA to read from the ADC FIFO and write to our buffer.
+    channel_config_set_transfer_data_size(&cfg, DMA_SIZE_16);
+    channel_config_set_read_increment(&cfg, false); // Read from the same address (ADC FIFO)
+    channel_config_set_write_increment(&cfg, true); // Write to sequential addresses into adcValues
+    channel_config_set_dreq(&cfg, DREQ_ADC);
+
+    /*
+        Setup the ring  buffer, 2^X so need 1, 2 and 4 positions
+
+        channelCount results in:
+        1 - 1 channel, no padding required
+        2 - 2 channels, no padding required
+        3 - 4 channels needed (due to ring buffer) so padding required if only 3 channels active
+        4 - 4 channels available - no padding required.
+
+        size_bits is the number of bytes = (1 << sizebits)
+        1 channel  => size bits = 1 (2 byte ring required i.e. 1 << 1)
+        2 channels => size bits = 2 (4 byte ring required i.e. 1 << 2)
+        4 channels => size bits = 4 (8 byte ring required i.e. 1 << 4)
+
+        We will never have 3 channels as 6 bytes is unavailable in the ring buffer,
+        hence the padding with the internal temp sensor to prevent misalignment.
+    */
+    const uint16_t sizeBits = llog2(popcount(mask)) * sizeof(uint16_t);
```

Comment:
sizeBits i number of bits to wrap - number of bytes is 1<<size_bits
```suggestion
    const uint16_t sizeBits = llog2(popcount(mask)) + sizeof(uint16_t) / 8 - 1;
```